services/web/server/src/simcore_service_webserver/exception_handling_base.py

```diff
@@ -0,0 +1,167 @@
+import functools
+import logging
+from collections.abc import Iterable
+from contextlib import AbstractAsyncContextManager
+from types import TracebackType
+from typing import Protocol, TypeAlias
+
+from aiohttp import web
+from servicelib.aiohttp.typing_extension import Handler as WebHandler
+from servicelib.aiohttp.typing_extension import Middleware as WebMiddleware
+
+_logger = logging.getLogger(__name__)
+
+
+class AiohttpExceptionHandler(Protocol):
+    __name__: str
+
+    async def __call__(
+        self,
+        request: web.Request,
+        exception: BaseException,
+    ) -> web.Response:
+        """
+        Callback that handles an exception produced during a request and transforms it into a response
+
+        Arguments:
+            request -- current request
+            exception -- exception raised in web handler during this request
+        """
+        ...  # pylint: disable=unnecessary-ellipsis
+
+
+ExceptionHandlersMap: TypeAlias = dict[type[BaseException], AiohttpExceptionHandler]
+
+
+def _sort_exceptions_by_specificity(
+    exceptions: Iterable[type[BaseException]], *, concrete_first: bool = True
```

Comment:
what is meant by concrete in this context?

Reply:
specificity refers to haw specific a class is. E.g. if you have classes A, B and C and

class A:
   ...
class B(A)
   ...
class C(B)
   ...


then the sort will return
[A, B, C] if `concrete_first = False`
[C, B, A] if `concrete_first = True`